### PR TITLE
Improve parser error recovery around misplaced semicolon

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/52

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler could i get eyes on this?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -2798,16 +2798,20 @@ private bool IsMisplacedModifier(SyntaxListBuilder modifiers, SyntaxList<Attribu
...
-                    result = this.ParsePropertyDeclaration(attributes, modifiers, type, explicitInterfaceOpt, identifierOrThisOpt, typeParameterListOpt);
-                    return true;
+                result = this.ParsePropertyDeclaration(attributes, modifiers, type, explicitInterfaceOpt, identifierOrThisOpt, typeParameterListOpt);
+                return true;
             }
 
             result = null;
             return false;
+
+            bool isStartOfPropertyBody(SyntaxKind kind)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:2813]: Consider extracting this and using below as well.

**Review by 333fred — COMMENTED** [review]
Done review pass (commit 3)

**Comment by CyrusNajmabadi** [comment]
@333fred done!  @cston ptal :)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -3851,10 +3866,14 @@ private SyntaxToken TryEatCheckedOrHandleUnchecked(ref SyntaxToken operatorKeywo
...
-            Debug.Assert(this.CurrentToken.Kind == SyntaxKind.EqualsGreaterThanToken ||
-                         this.CurrentToken.Kind == SyntaxKind.OpenBraceToken);
+            // Error recovery: add an errant semicolon to the identifier token and keep going.
+            if (this.CurrentToken.Kind is SyntaxKind.SemicolonToken)
+            {
+                identifier = AddTrailingSkippedSyntax(identifier, this.EatTokenWithPrejudice(SyntaxKind.OpenBraceToken));
+            }
+
+            // We know we are parsing a property because we have seen either an open brace or an arrow token
+            Debug.Assert(this.CurrentToken.Kind is SyntaxKind.EqualsGreaterThanToken or SyntaxKind.OpenBraceToken);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:3876]: Here too

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Whoops.  Fixed now :)

**Review by 333fred — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@jcouv @cston could you ptal?  thanks!

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -2798,18 +2798,22 @@ private bool IsMisplacedModifier(SyntaxListBuilder modifiers, SyntaxList<Attribu
                 return true;
             }
 
-            switch (this.CurrentToken.Kind)
+            // `{` or `=>` definitely start a property.  Also allow
+            // `; {` and `; =>` as error recovery fo ra misplaced semicolon.
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:2802]: >fo ra [](http://example.com/codeflow?start=50&length=5)

Typo.

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:2802]:
```suggestion
            // `; {` and `; =>` as error recovery for a misplaced semicolon.
```

File: src/Compilers/CSharp/Test/Syntax/Parsing/MemberDeclarationParsingTests.cs [hunk 4]
```diff
@@ -9107,5 +9107,107 @@ public void ReadonlyParameter4()
...
+                            }
+                        }
+                        N(SyntaxKind.SemicolonToken);
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+            EOF();
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/MemberDeclarationParsingTests.cs:9211]: > } [](http://example.com/codeflow?start=7&length=2)

Are we testing script cases for global statements?

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/Parsing/MemberDeclarationParsingTests.cs:9211]: we are, yes.  and they def discovered interesting stuff for example, that `int MyProperty; {` is totally legal at the script/top-level as a local/field, followed by a block :)

**Pull request merged**